search-app/src/cmr/search/api/concepts_search.clj

```diff
@@ -47,6 +53,34 @@
         results (query-svc/find-concepts-by-json-query ctx concept-type params json-query)]
     (core-api/search-response ctx results)))
 
+(defn- all-granule-params?
+  "Check if the params is all granule params based on the concept-type,query params and scroll-id.
+   Only granule queries that don't have scroll-id and don't contain certain collection constraints
+   are all granule params."
+  [concept-type params scroll-id]
+  (let [constraints (select-keys 
+                      params 
+                      all-gran-validation/granule-limiting-search-fields)]
+    (and (= :granule concept-type)
+         (not (some? scroll-id))
+         (not (seq (util/remove-nil-keys constraints))))))
+
+(defn- handle-all-granule-params
+  "Throws error if all granule params needs to be rejected."
+  [headers]
+  (when (and (= false (allow-all-granule-params-flag))
+             (or (not (some? (get headers "client-id")))
+                 (not (some? (get headers "allow-all-gran-query")))))
+    (svc-errors/throw-service-error
+      :bad-request
+      (str "The CMR does not currently allow querying across granules in all collections. "
+          "To help optimize your search, you should limit your query using conditions that "
+          "identify one or more collections, such as provider, provider_id, concept_id, "
+          "collection_concept_id, short_name, version, entry_title or entry_id. "
+          "For further support request, please contact support@earthdata.nasa.gov, "
```

Comment:
Visit the CMR Client Developer Forum at https://wiki.earthdata.nasa.gov/display/CMR/Granule+Queries+Now+Require+Collection+Identifiers for more information, and for any questions please contact support@earthdata.nasa.gov.